packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx

```diff
@@ -180,14 +208,99 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
     });
   }
 
+  // Get indices of selected elements in options array when we have selected values.
+  const getSelectedIndexes = React.useCallback(() => {
+    const indexArray: number[] = [];
+    if (isMultipleEnabled(valueProp, multiple) && !!valueProp) {
+      valueProp.forEach((value) => {
+        indexArray.push(options.findIndex((option) => option.value === value));
+      });
+    }
+    return indexArray;
+  }, [multiple, options, valueProp]);
+
   // Reducer where all the component-wide state is stored
+  const comboBoxReducer = (
+    state: {
+      isOpen: boolean;
+      selectedIndex: number | number[];
+      focusedIndex: number;
+    },
+    {
+      type,
+      value,
+    }:
+      | { type: 'multiselect'; value: number }
+      | { type: 'multi-override'; value: number[] }
+      | { type: 'open'; value: undefined }
+      | { type: 'close'; value: undefined }
+      | { type: 'select'; value: number }
+      | { type: 'focus'; value: number | undefined },
+  ) => {
+    switch (type) {
+      case 'open': {
+        return { ...state, isOpen: true };
+      }
+      case 'close': {
+        return { ...state, isOpen: false };
+      }
+      case 'select': {
+        if (isSingleOnChange(onChangeProp.current, multiple)) {
+          onChangeProp.current?.(optionsRef.current[value]?.value);
+        }
+        return {
+          ...state,
+          selectedIndex: value ?? state.selectedIndex,
+          focusedIndex: value ?? state.focusedIndex,
+        };
+      }
+      case 'multi-override': {
+        return { ...state, selectedIndex: value };
+      }
+      case 'multiselect': {
+        const actionIndex = (state.selectedIndex as number[]).findIndex(
```

Comment:
I don't think we should be type casting to `as number[]` everywhere without actually checking the type, because then we are *lying* to both typescript and ourselves.

A better strategy would be to check using `Array.isArray` and early return/throw if that's false.